File: aquatrack_backend/app/Support/ApiResponse.php
<?php

namespace App\Support;

use Illuminate\Http\JsonResponse;

trait ApiResponse
{
    protected function successResponse(mixed $data = null, string $message = 'OK', int $code = 200): JsonResponse
    {
        return response()->json([
            'success' => true,
            'data' => $data,
            'message' => $message,
        ], $code);
    }

    protected function errorResponse(string $message = 'Error', mixed $data = null, int $code = 400): JsonResponse
    {
        return response()->json([
            'success' => false,
            'data' => $data,
            'message' => $message,
        ], $code);
    }
}

File: aquatrack_mobile/lib/data/models/api_response.dart
class ApiResponse<T> {
  ApiResponse({
    required this.success,
    required this.data,
    required this.message,
  });

  final bool success;
  final T data;
  final String message;

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic rawData) mapData,
  ) {
    return ApiResponse<T>(
      success: json['success'] as bool? ?? false,
      data: mapData(json['data']),
      message: json['message'] as String? ?? '',
    );
  }
}
